minigrid/envs/babyai/core/levelgen.py

```diff
@@ -0,0 +1,478 @@
+"""
+Copied and adapted from https://github.com/mila-iqia/babyai
+"""
+from minigrid.core.constants import COLOR_NAMES
+from minigrid.core.roomgrid import Room, RoomGrid
+from minigrid.minigrid_env import MissionSpace
+
+from .verifier import (
+    LOC_NAMES,
+    OBJ_TYPES,
+    OBJ_TYPES_NOT_DOOR,
+    ActionInstr,
+    AfterInstr,
+    AndInstr,
+    BeforeInstr,
+    GoToInstr,
+    ObjDesc,
+    OpenInstr,
+    PickupInstr,
+    PutNextInstr,
+    SeqInstr,
+)
+
+
+class RejectSampling(Exception):
+    """
+    Exception used for rejection sampling
+    """
+
+    pass
+
+
+class BabyAIMissionSpace(MissionSpace):
+    """
+    Class that mimics the behavior required by minigrid.minigrid_env.MissionSpace,
+    but does not change how missions are generated for BabyAI. It silences
+    the gymnasium.utils.passive_env_checker given that it considers all strings to be
+    plausible samples.
+    """
+
+    def __init__(self):
+        super().__init__(mission_func=lambda: "go")
+
+    def contains(self, x: str):
+        return True
+
+
+class RoomGridLevel(RoomGrid):
```

Comment:
Can this be split out into a separate file as it seems like a separate object to level gen

Reply:
RoomGridLevel and LevelGen play similar roles (All BabyAI levels/environments inherit from one of them). LevelGen allows for the more complex instructions (those that chain simple instructions). Maybe we should just rename the file to something else.

Reply:
The file is currently massive with two large classes in it
It would make it easier to read if we moved the two classes into separate files if possible.